### DISCUSSION TITLE
Help with black screen on IPad

### DISCUSSION THREAD

**Discussion by FOOincognita** [coder/code-server#2753] [Q&A]
- Web Browser: Chrome, Safari 
- Local OS: IPadOS 14.4, IPad Air 4th gen.
- Remote OS: Ubuntu 16.04, Ubuntu 18.04, Ubuntu 20.04 (I have tried all 3)
- Remote Architecture: AMD64
- `code-server --version`:  
- 1.408-vsc1.32.0 with Ubuntu 16.04
-  3.4.0 & 3.9.0 with Ubuntu 18.04 & Ubuntu 20.04 (all combinations)
- Google Cloud Platform VM Instance (32GB SSD, 4 cores, 8GB RAM, HTTPS/HTTP traffic enabled)

Howdy! I am a university student, and Code-Server would be absoluetly perfect for me so that I could have a remote instance of VS Code for when I am away from my main PC. Unfortunately, I can only gte it to work on other PCs, and not my IPad. If I could get it working on my IPad, it would seriously be a life-saver. 

Please keep in mind I have incredibly little experience with linux besides C++ compilation through WSL, and Code-server basic install and run, but the instructions on this GitHub for authentication or SSH are pure gibberish to me, so maybe that's my issue? 

I have looked all over YouTube, this GitHub, message boards, and more, but I can't seem to resolve the black screen issue (allows me to enter password, but then goes to black screen with webpage fully loaded). Code-Server is runny seamlessly for any of my PC browsers, though, whenever I try to access the VM through my IPad on any browser, I get the same issue; It will not show anything other than a black screen after correct password entry. 

I have seen other say its an HTTPS issue, something about authentication errors specific to the IPad, but I honestly only have experience with C++, and that's honesty it; I have no clue how to go about fixing that.

Now, I found an app called "Servereditor for code-server" that allows you to pay for a pre-built instance, or allows you to connect to a self-host. The paid server connects just fine, but it is missing the most important feature; access to bash for code compilation and page installs. When I try to connect to my self-hosted server via the URL, it brings up the password screen, then just says "Invalid Password", even if I copy and paste the password directly from cloud console...I am out of ideas and was hoping maybe I missed a post somewhere where someone found a way to fix it?

Please let me know, and thank you for your time!

**Comment by jsjoeio** [comment]
Hey @FOOincognita - sorry to hear you've run into some issues with code-server on iPad!

We would love to help you figure this out. I'm going to move this over to our Discussions page.

**Reply by jsjoeio** [reply]
There's a few things we can try:
- can you try running code-server with `--log debug`? More details [here](https://github.com/cdr/code-server/blob/v3.9.0/docs/FAQ.md#how-do-i-debug-issues-with-code-server) and then post the logs here. 
- there could be errors in the browser that might help. unfortunately the only app i know that gives you that is paid $6.99 called [Inspect Browser](https://apps.apple.com/us/app/inspect-browser/id1203594958). there may be other solutions to get access to the logs but you'd have to google it
- if you provide reproduction steps for how you set up code-server with GCP (I haven't gone through this yet), I can try doing that myself and seeing if I get the same error

**Reply by TheGreatestJannet** [reply]
Just to let other people know if you have an iMac you can connect the iPad to the iMac and use the developer tools in safari to get the console from the iPad.

**Comment by TheGreatestJannet** [comment]
I have the same issue as OP on an iPad Air 2 with a letsencrypt HTTPS certificate. I’m also running iPadOS 14.4. I’m using the linuxserver docker image as my server is running on ARM64. Any idea how I would enable logging there?

**Reply by jsjoeio** [reply]
You should be able to run code-server with `--log debug`. More info [here](https://github.com/cdr/code-server/blob/v3.9.0/docs/FAQ.md#how-do-i-debug-issues-with-code-server)

**Reply by TheGreatestJannet** [reply]
Here is the error: 
![CBC2BDD2-7DCF-4E29-9786-DFDA1723ACD2](https://user-images.githubusercontent.com/51300636/110773905-259fe900-8255-11eb-8c14-18515f2f8eb8.jpeg)

**Reply by jsjoeio** [reply]
Thanks for the screenshot! Interesting...Something in the `workbench.web.api.js` that's causing a DOMException 🤔

If it's possible, do you mind providing reproduction steps? Ideally starting from how you set up code-server on the remote machine? I could try to create the same setup and reproduce.

**Reply by jsjoeio** [reply]
The other thing i'm seeing..."navigator is undefined".

I wrote a test and fixed that in this PR https://github.com/cdr/code-server/pull/2719 which was released in 3.9.1.

Can you try upgrading code-server and seeing if the issue still persists?

**Reply by TheGreatestJannet** [reply]
> Thanks for the screenshot! Interesting...Something in the `workbench.web.api.js` that's causing a DOMException thinking
> 
> If it's possible, do you mind providing reproduction steps? Ideally starting from how you set up code-server on the remote machine? I could try to create the same setup and reproduce.

Here is my docker-compose:
```
version: "2.1"
services:
  code-server:
    image: linuxserver/code-server
    container_name: code-server
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/London
      - PASSWORD=password
      - HASHED_PASSWORD=password
      - SUDO_PASSWORD=password #optional
      - PROXY_DOMAIN=code-server.server.org
      - DOCKER_MODS=linuxserver/mods:code-server-shellcheck|linuxserver/mods:code-server-python3|linuxserver/mods:code-server-php
    volumes:
      - /home/user/code-server/config:/config
      - /var/www:/livewebsite
    ports:
      - 8445:8443
    restart: unless-stopped

```
and here is my nginx file:
```
server{
        listen 80;
        server_name code-server.server.org;

        # Uncomment to redirect HTTP to HTTPS
        return 301 https://$host$request_uri;
}

server {
    listen 443;
    listen [::]:443;
    server_name code-server.server.org;

	add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
        add_header Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-inline'; connect-src 'self' extensions.coder.com; img-src 'self' raw.githubusercontent.com dev.azure.com storage.googleapis.com vsmarketplacebadge.apphb.com github.com  data:; style-src 'self' 'unsafe-inline'; font-src 'self';" always;
        add_header X-XSS-Protection "1; mode=block";
        add_header X-Frame-Options "SAMEORIGIN";
        add_header X-Content-Type-Options nosniff;
        add_header Referrer-Policy "strict-origin";
        add_header Permissions-Policy "geolocation=(),midi=(),sync-xhr=(),microphone=(),camera=(),magnetometer=(),gyroscope=(),fullscreen=(self),payment=()";
	add_header Expect-CT "enforce, max-age=3600";

    location / {
      auth_basic "Restricted Content";
      auth_basic_user_file /etc/nginx/.htpasswd;
      proxy_pass http://127.0.0.1:8445/;
      proxy_set_header Host $host;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection upgrade;
      proxy_set_header Accept-Encoding gzip;
    }

    ssl_certificate /etc/nginx/certs/wildcard.server.org/fullchain.pem;
    ssl_certificate_key /etc/nginx/certs/wildcard.server.org/key.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
```

> The other thing i'm seeing..."navigator is undefined".
> 
> I wrote a test and fixed that in this PR #2719 which was released in 3.9.1.
> 
> Can you try upgrading code-server and seeing if the issue still persists?

I am running the latest version of code-server so I should already have the patch.

**Reply by TheGreatestJannet** [reply]
here is another console log:
```
[Error] Failed to load resource: the server responded with a status of 404 () (vscode.js.map, line 0)
[Error] WebSocket connection to 'wss://code-server.server.org/?reconnectionToken=b56463a7-f40e-4d3c-b469-e916d7ab6eb4&reconnection=false&skipWebSocketFrames=false' failed: Unexpected response code: 401
[Error] Unhandled Promise Rejection: [object Event]
	(anonymous function)
	promiseReactionJob
[Error] Unhandled Promise Rejection: Error: Assertion Failed: argument is undefined or null
	(anonymous function) (workbench.web.api.js:56:439896)
	promiseReactionJob
[Error] [object Event]
	(anonymous function) (workbench.web.api.js:5)
[Error] Unhandled Promise Rejection: [object Event]
	(anonymous function)
	promiseReactionJob
[Error] Error: Unable to read file 'vscode-remote://code-server.server.org/config/workspace/.vscode/settings.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote://code-server.server.org/config/workspace/.vscode/tasks.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote://code-server.server.org/config/workspace/.vscode/launch.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote://code-server.server.org/config/workspace/.vscode/settings.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote://code-server.server.org/config/workspace/.vscode/tasks.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote://code-server.server.org/config/workspace/.vscode/launch.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote:/config/data/User/settings.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
[Error] Error: Unable to read file 'vscode-remote:/config/data/User/tasks.json' ([object Event])

w@https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:612:1845160
https://code-server.server.org/static/e0203f2a36c9b7036fefa50eec6cf8fa36c5c015/usr/local/share/.config/yarn/global/node_modules/code-server/lib/vscode/out/vs/workbench/workbench.web.api.js:616:1915359
asyncFunctionResume@[native code]
[native code]
promiseReactionJobWithoutPromise@[native code]
	(anonymous function) (workbench.web.api.js:5)
```
It works fine on my desktop and Firefox.

**Reply by jsjoeio** [reply]
Hmm..thanks for the logs! Yeah seems like we'll need to look into this a bit more

**Reply by TheGreatestJannet** [reply]
It seems it is trying to setup websockets over port 80 and not 443 which causing it to blocked on Safari but not on Firefox. This might be part of the issue.

**Reply by jsjoeio** [reply]
Maybe @code-asher might know something that could help?

**Reply by TheGreatestJannet** [reply]
I’m not sure if it is because of my CSP that it is being blocked. But the text log above is with CSP disabled so that isn’t the root cause. I might be wrong but isn’t websockets over port 80 insecure anyway so it should use 443 regardless?

**Reply by code-asher** [reply]
The 80 is a red herring because we patch VS Code to use the current port (so 443 in this case) but we do it at the point where the websocket is created so the logs erroneously still report 80. We should fix that since it's confusing.

I find the 401 (unauthorized) code particularly interesting:

```
WebSocket connection to 'wss://code-server.server.org/[...]' failed: Unexpected response code: 401
```

Are you using code-server's password authentication? Could you check the network tab and see if a cookie with `key` is being sent in the request headers on the websocket request?

**Reply by TheGreatestJannet** [reply]
I don’t seem to have a websocket request.
![BFFD051E-7E3B-4A4A-85AD-C3FBA02BAC7E](https://user-images.githubusercontent.com/51300636/111066374-9a259280-84b6-11eb-9063-5c0270c45c7b.jpeg)
![C97A2168-49E3-47ED-B029-0F7803C52CEA](https://user-images.githubusercontent.com/51300636/111066375-9c87ec80-84b6-11eb-84ac-c57c24bd8793.jpeg)

**Reply by TheGreatestJannet** [reply]
I’ve solved it! It seems that using nginx auth basic breaks code server on iPad and so does my CSP for some reason even though both work perfectly on every other browser...

**Reply by TheGreatestJannet** [reply]
Here is my current nginx config that works perfectly on Safari and Firefox:
```
server{
        listen 80;
        server_name code-server.server.org;

        # Uncomment to redirect HTTP to HTTPS
        return 301 https://$host$request_uri;
}

server {
    listen 443;
    listen [::]:443;
    server_name code-server.server.org;

        add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
        add_header Content-Security-Policy "default-src 'self'; script-src 'self' 'unsafe-eval' 'unsafe-inline'; connect-src 'self' wss://code-server.server.org extensions.coder.com; img-src 'self' raw.githubusercontent.com dev.$
        add_header X-XSS-Protection "1; mode=block";
        add_header X-Frame-Options "SAMEORIGIN";
        add_header X-Content-Type-Options nosniff;
        add_header Referrer-Policy "strict-origin";
        add_header Permissions-Policy "geolocation=(),midi=(),sync-xhr=(),microphone=(),camera=(),magnetometer=(),gyroscope=(),fullscreen=(self),payment=()";
        add_header Expect-CT "enforce, max-age=3600";

    location / {
      proxy_pass http://127.0.0.1:8445/;
      proxy_set_header Host $host;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection upgrade;
      proxy_set_header Accept-Encoding gzip;
    }

    #auth_basic "Restricted Content";
    #auth_basic_user_file /etc/nginx/.htpasswd;

    ssl_certificate /etc/nginx/certs/wildcard.server.org/fullchain.pem;
    ssl_certificate_key /etc/nginx/certs/wildcard.server.org/key.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
}
```
Auth basic breaking websockets seems to be a Safari limitation: https://github.com/cdr/code-server/issues/1348.

Unsafe-eval seems required to get text colouring in Safari and Chrome. Is there a way to prevent this? It seems like a pretty big security risk to allow it. Firefox doesn't seem to require it.

Finally, I can't seem to get the onscreen keyboard to show up on my iPad. If it does pop up it seems to move the whole screen up making it very hard or impossible to see what you are typing. Should I open a new issue for this? Thank you very much for your help!

**Reply by jsjoeio** [reply]
> I can't seem to get the onscreen keyboard to show up on my iPad. If it does pop up it seems to move the whole screen up making it very hard or impossible to see what you are typing

Can you take a look at #979? There might be a helpful tip in there. Although I will say, based on my experience using code-server on an iPad, an external keyboard is the best way to use it.

**Reply by code-asher** [reply]
Nice find on the basic auth issue!

I'm not sure about unsafe-eval, might be something we should look into.
